be/src/exec/pipeline/sink/file_sink_operator.cpp

```diff
@@ -0,0 +1,300 @@
+
+    std::atomic_int32_t _num_result_sinkers = 0;
+    std::atomic_int64_t _num_pending_chunks = 0;
+
+    bool _is_writer_opened = false;
+    std::atomic_bool _is_prepared = false;
+    std::atomic_bool _is_cancelled = false;
+    std::atomic_bool _is_finished = false;
+
+    mutable SpinLock _io_status_mutex;
+    Status _io_status;
+
+    FragmentContext* const _fragment_ctx;
+    RuntimeState* _state = nullptr;
+
+    static const int32_t kExecutionQueueSizeLimit = 64;
+};
+
+// this function will be executed in bthread
+int FileSinkBuffer::execute_io_task(void* meta, bthread::TaskIterator<const vectorized::ChunkPtr>& iter) {
+    FileSinkBuffer* file_sink_buffer = static_cast<FileSinkBuffer*>(meta);
+    for (; iter; ++iter) {
+        file_sink_buffer->_process_chunk(iter);
+    }
+    return 0;
+}
+
+void FileSinkBuffer::_process_chunk(bthread::TaskIterator<const vectorized::ChunkPtr>& iter) {
+    --_num_pending_chunks;
+    // close is already done, just skip
+    if (_is_finished) {
+        return;
+    }
+
+    // cancelling has happened but close is not invoked
+    if (_is_cancelled && !_is_finished) {
+        close(_state);
+        return;
+    }
+
+    if (!_is_writer_opened) {
+        if (Status status = _process_writer_op([this]() { return _writer->open(_state); }); !status.ok()) {
+            set_io_status(status);
+            close(_state);
+            return;
+        }
+        _is_writer_opened = true;
+    }
+    auto chunk = *iter;
+    if (chunk == nullptr) {
+        // this is the last chunk
+        close(_state);
+        return;
+    }
+    if (Status status = _process_writer_op([this, chunk]() { return _writer->append_chunk(chunk.get()); });
+        !status.ok()) {
+        set_io_status(status);
+        close(_state);
+    }
+}
+
+Status FileSinkBuffer::_process_writer_op(std::function<Status()> op) {
+    // for some writers(eg. hdfs) that cannot be executed in bthread, we need to put them in pthread.
+    if (_writer->can_run_in_bthread()) {
+        return op();
+    } else {
+        auto ret = call_hdfs_scan_function_in_pthread(op);
+        return ret->get_future().get();
+    }
+}
+
+Status FileSinkBuffer::prepare(RuntimeState* state, RuntimeProfile* parent_profile) {
+    bool expected = false;
+    if (!_is_prepared.compare_exchange_strong(expected, true)) {
+        return Status::OK();
+    }
+
+    RETURN_IF_ERROR(state->exec_env()->result_mgr()->create_sender(state->fragment_instance_id(), 1024, &_sender));
+
+    _state = state;
+    _writer = std::make_shared<FileResultWriter>(_file_opts.get(), _output_expr_ctxs, parent_profile);
+    RETURN_IF_ERROR(_writer->init(state));
+
+    bthread::ExecutionQueueOptions options;
+    _exec_queue_id = std::make_unique<bthread::ExecutionQueueId<const vectorized::ChunkPtr>>();
+    int ret = bthread::execution_queue_start<const vectorized::ChunkPtr>(_exec_queue_id.get(), &options,
```

Comment:
Why here using bthread instead of io threads the same as the `OlapScanOperator`? In this way, maybe we cannot apply resource group feature to `FileSink`.